### ISSUE TITLE
【论文复现】反向梯度无法对齐

### ISSUE THREAD

**Issue by inspiration12138** [PaddlePaddle/Paddle#45825]
### 请提出你的问题 Please ask your question

您好，在论文复现的过程中已经实现模型的正向对齐和损失函数对齐，反向从第二轮损失开始对不齐，通过打印torch和paddle的反向梯度，通过ReprodLogger进行diff验证，发现是使用的paddle.vision.resnet的预训练模型中的downsample层开始diff到了e-4级别，在这下面的梯度均为e-5及以下，请问这种情况下有什么解决方案吗

附上主要的测试代码
        import sys
        import os
        import argparse
        import random
        
        import torch
        import paddle
        import numpy as np
        from reprod_log import ReprodDiffHelper
        from reprod_log import ReprodLogger
        
        seed = 456
        #
        paddle.seed(seed)
        np.random.seed(seed)
        random.seed(seed)
        # os.environ['PYTHONHASHSEED'] = str(seed)
        torch.manual_seed(seed)
        # torch.cuda.manual_seed(seed)
        # torch.cuda.manual_seed_all(seed)
        
        from FCFR_Net_ref.model import DepthCompletionNet as DepthCompletionNet_torch
        from FCFR_Net_paddle.model import DepthCompletionNet as DepthCompletionNet_paddle
        import FCFR_Net_paddle.criteria as criteria_paddle
        import FCFR_Net_ref.criteria as criteria_torch
        
        args = parser.parse_args()
        args.use_pose = ("photo" in args.train_mode)
        # args.pretrained = not args.no_pretrained
        args.result = os.path.join('..', 'results')
        args.use_rgb = ('rgb' in args.input) or args.use_pose
        args.use_d = 'd' in args.input
        args.use_g = 'g' in args.input
        if args.use_pose:
            args.w1, args.w2 = 0.1, 0.1
        else:
            args.w1, args.w2 = 0, 0
        print(args)
        
        def getmodel(TYPE):
            global args
        
            if TYPE == 'paddle':
                # load paddle model
                paddle.set_device("gpu")
                paddle_model = DepthCompletionNet_paddle(args)
                paddle_state_dict = paddle.load("FCFR_Net_paddle/model_best.pdparams")
                paddle_model.set_dict(paddle_state_dict)
                paddle_model.eval()
                return paddle_model
            else:
                # load torch model
                torch_model = DepthCompletionNet_torch(args).cuda()
                torch_state_dict = torch.load("FCFR_Net_ref/model_best.pth.tar")['model']
                torch_model.load_state_dict(torch_state_dict)
                torch_model.eval()
                return torch_model
        
        def getdata(TYPE):
            # Data loading code
            d = np.load("data/fake_d.npy")
            rgb = np.load("data/fake_rgb.npy")
            g = np.load("data/fake_g.npy")
            gt = np.load("data/fake_gt.npy")
            if TYPE == 'torch':
                inputs_torch = {"rgb": torch.tensor(rgb,device=device), "d": torch.tensor(d,device=device),
                                "g": torch.tensor(g,device=device),"gt": torch.tensor(gt,device=device)}
                return inputs_torch
            else:
                inputs_paddle = {"rgb": paddle.to_tensor(rgb), "d": paddle.to_tensor(d), "g": paddle.to_tensor(g), "gt": paddle.to_tensor(gt)}
                return inputs_paddle
        
        
        def train_one_epoch(TYPE,inputs, model, criterion, optimizer, max_iter, reprod_logger):
            print('start test training...')
            model.eval()
        
            for idx in range(max_iter):
                print('==> start idx {} ...'.format(idx))
                pred = model(inputs)
                depth_loss = criterion(pred, inputs['gt'])  # +depth_criterion(global_features, res_gt)
                reprod_logger.add("loss_{}".format(idx), depth_loss.cpu().detach().numpy())
        
                if TYPE == 'torch':
                    print('torch_{} ==> {}'.format(idx,depth_loss))
                    optimizer.zero_grad()
                    depth_loss.backward()
        
                    for item in model.named_parameters():
                        name,param = item
                        grad = param.grad
                        reprod_logger.add("times_{}_name_{}".format(idx,name), grad.data.cpu().detach().numpy())
        
                    optimizer.step()
                else:
                    print('paddle_{} ==> {}'.format(idx, depth_loss))
                    optimizer.clear_grad()
                    depth_loss.backward()
        
                    for item in model.named_parameters():
                        name,param = item
                        grad = param.grad
                        if name.endswith("weight") or name.endswith('bias'):
                            reprod_logger.add("times_{}_name_{}".format(idx,name), grad.cpu().detach().numpy())
        
                    optimizer.step()
        
            if TYPE == 'torch':
                reprod_logger.save("result/losses_ref.npy")
            else:
                reprod_logger.save("result/losses_paddle.npy")
        
        
        def test_backward(TYPE):
            global args
            max_iter = 1
        
            model = getmodel(TYPE)
            inputs = getdata(TYPE)
            reprod_logger = ReprodLogger()
        
            # init optimizer and criterion
            if TYPE == 'torch':
                model_named_params = [p for _, p in model.named_parameters() if p.requires_grad]
                optimizer = torch.optim.Adam(params=model_named_params, lr=args.lr,betas=(0.9, 0.999), eps=1e-8,weight_decay=0.0001)
                criterion = criteria_torch.MaskedMSELoss()# if (args.criterion == 'l2') else criteria_torch.MaskedL1Loss()
            else:
                model_named_params = [p for _, p in model.named_parameters() if not p.stop_gradient]
                optimizer = paddle.optimizer.Adam(learning_rate=args.lr, parameters=model_named_params,beta1=0.9,beta2=0.999, epsilon=1e-8,weight_decay=0.0001)
                criterion = criteria_paddle.MaskedMSELoss()# if (args.criterion == 'l2') else criteria_paddle.MaskedL1Loss()
        
            train_one_epoch(TYPE, inputs, model, criterion, optimizer, max_iter, reprod_logger)
        
        if __name__ == '__main__':
            # TYPE = 'paddle'
            device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
            test_backward('torch')
            torch.cuda.empty_cache()
            test_backward('paddle')
        
            # load data
            diff_helper = ReprodDiffHelper()
            torch_info = diff_helper.load_info("result/losses_ref.npy")
            paddle_info = diff_helper.load_info("result/losses_paddle.npy")
        
            # compare result and produce log
            diff_helper.compare_info(torch_info, paddle_info)
            diff_helper.report(path="result/log/backward_diff.log")

**Comment by LielinJiang** [comment]
误差是会累计的，e-5经过一次计算变长e-4是非常正常的情况。可以从结果来看下是否能够对齐

**Comment by inspiration12138** [comment]
从结果来看就是对不齐，误差都在e-3左右
附上diff结果
        [2022/09/07 18:58:04] root INFO: loss_0: 
        [2022/09/07 18:58:04] root INFO: 	mean diff: check passed: True, value: 0.0
        [2022/09/07 18:58:04] root INFO: loss_1: 
        [2022/09/07 18:58:04] root INFO: 	mean diff: check passed: False, value: 0.001953125
        [2022/09/07 18:58:04] root INFO: loss_2: 
        [2022/09/07 18:58:04] root INFO: 	mean diff: check passed: False, value: 0.006591796875
        [2022/09/07 18:58:04] root INFO: loss_3: 
        [2022/09/07 18:58:04] root INFO: 	mean diff: check passed: False, value: 0.005126953125
        [2022/09/07 18:58:04] root INFO: loss_4: 
        [2022/09/07 18:58:04] root INFO: 	mean diff: check passed: False, value: 0.000244140625
        [2022/09/07 18:58:04] root INFO: diff check failed

**Comment by inspiration12138** [comment]
在paddle运行的时候有warning，但是不影响运行，用paddle.utilities.run_check()也是通过的，这个是否会影响到计算呢？
W0908 11:18:46.313169 18668 gpu_resources.cc:61] Please NOTE: device: 0, GPU Compute Capability: 8.6, Driver API Version: 11.6, Runtime API Version: 11.1
W0908 11:18:46.314168 18668 gpu_resources.cc:91] device: 0, cuDNN Version: 8.0.
W0908 11:18:51.017967 18668 gpu_resources.cc:201] WARNING: device:  . The installed Paddle is compiled with CUDNN 8.1, but CUDNN version in your machine is 8.0, which may cause serious incompatible bug. Please recompile or reinstall Paddle with compatible CUDNN version.

**Comment by LielinJiang** [comment]
这个diff是loss的diff嘛？我说的结果是最终的训练的精度

**Comment by inspiration12138** [comment]
最终训练的精度用小数据集跑可以达到，但是要求torch和paddle的模型在loss上要对齐
